pkg/resource/archive/hooks.go

```diff
@@ -0,0 +1,99 @@
+package archive
+
+import (
+	"errors"
+	"fmt"
+
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
+	"github.com/aws/aws-sdk-go/service/eventbridge"
+
+	"github.com/aws-controllers-k8s/eventbridge-controller/apis/v1alpha1"
+)
+
+const (
+	StatusEnabled      = "ENABLED"
+	StatusDisabled     = "DISABLED"
+	StatusCreating     = "CREATING"
+	StatusUpdating     = "UPDATING"
+	StatusCreateFailed = "CREATE_FAILED"
+	StatusUpdateFailed = "UPDATE_FAILED"
+)
+
+// TerminalStatuses are the status strings that are terminal states for an
+// Archive
+var TerminalStatuses = []string{
+	StatusCreateFailed,
+	StatusUpdateFailed,
+}
+
+// archiveInTerminalState returns whether the supplied Archive is in a terminal
+// state
+func archiveInTerminalState(r *resource) bool {
+	if r.ko.Status.State == nil {
+		return false
+	}
+	state := *r.ko.Status.State
+	for _, s := range TerminalStatuses {
+		if state == s {
+			return true
+		}
+	}
+	return false
+}
+
+// archiveAvailable returns true if the supplied Archive is in an available
+// status
+func archiveAvailable(r *resource) bool {
+	if r.ko.Status.State == nil {
+		return false
+	}
+	state := *r.ko.Status.State
+	return state == StatusEnabled || state == StatusDisabled
+}
+
+// archiveCreating returns true if the supplied Archive is in the process of
+// being created
+func archiveCreating(r *resource) bool {
+	if r.ko.Status.State == nil {
+		return false
+	}
+	state := *r.ko.Status.State
+	return state == StatusCreating
+}
+
+// requeueWaitUntilCanModify returns a `ackrequeue.RequeueNeededAfter` struct
+// explaining the Archive cannot be modified until it reaches an available
+// status.
+func requeueWaitUntilCanModify(r *resource) *ackrequeue.RequeueNeededAfter {
+	if r.ko.Status.State == nil {
+		return nil
+	}
+	status := *r.ko.Status.State
+	msg := fmt.Sprintf(
+		"Archive in '%s' state, cannot be modified.",
+		status,
+	)
+	return ackrequeue.NeededAfter(
+		errors.New(msg),
+		ackrequeue.DefaultRequeueAfterDuration,
+	)
+}
+
+// if an optional desired field value is nil explicitly unset it in the request
+// input
+func unsetRemovedSpecFields(
+	spec v1alpha1.ArchiveSpec,
+	input *eventbridge.UpdateArchiveInput,
+) {
+	if spec.EventPattern == nil {
+		input.SetEventPattern("")
+	}
```

Comment:
Note sure but.. should we wrap this with a `if delta.DifferentAt(`Spec.EventPattern`)` ?

Reply:
@A-Hilaly probably not necessary at this point since `unsetRemovedSpecFields` will only be called if one of these fields is different from desired. But, I suppose it wouldn't hurt.

Reply:
@A-Hilaly not sure if this covers case I added this which is during updates. 

This is the current code for `newUpdateRequestPayload`:

```go
func (rm *resourceManager) newUpdateRequestPayload(
	ctx context.Context,
	r *resource,
) (*svcsdk.UpdateArchiveInput, error) {
	res := &svcsdk.UpdateArchiveInput{}

	if r.ko.Spec.ArchiveName != nil {
		res.SetArchiveName(*r.ko.Spec.ArchiveName)
	}
	if r.ko.Spec.Description != nil {
		res.SetDescription(*r.ko.Spec.Description)
	}
	if r.ko.Spec.EventPattern != nil {
		res.SetEventPattern(*r.ko.Spec.EventPattern)
	}
	if r.ko.Spec.RetentionDays != nil {
		res.SetRetentionDays(*r.ko.Spec.RetentionDays)
	}

	return res, nil
}
```

If `Description` is `nil` during the update of another value, say `RetentionDays` , `delta` won't be useful here as the `UpdateArchiveWithContext` API requires values for all fields or will fail with validation exception during update (see validation patterns: https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_UpdateArchive.html)

Reply:
Haaaa, gotcha! Thank you for taking time to explain this to me :)